sql/distsqlrun: TestFlowRegistryDrain/MinFlowWait failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/07890702e2faabd34b7885470b141a5612d862cc

Parameters:
```
TAGS=
GOFLAGS=-race
```

Stress build found a failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=500025&tab=buildLog

```


goroutine 3352 [running]:
testing.tRunner.func1(0xc4215060f0)
	/usr/local/go/src/testing/testing.go:711 +0x5d9
runtime.Goexit()
	/usr/local/go/src/runtime/panic.go:366 +0x150
testing.(*common).FailNow(0xc421506780)
	/usr/local/go/src/testing/testing.go:544 +0x59
testing.(*common).Fatal(0xc421506780, 0xc420e19ef0, 0x1, 0x1)
	/usr/local/go/src/testing/testing.go:582 +0x7d
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.TestFlowRegistryDrain.func1(0x3a4f0c4ea73a2449, 0x80133bddabec4e9f)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow_registry_test.go:381 +0x172
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.TestFlowRegistryDrain.func5(0xc4215060f0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow_registry_test.go:451 +0xc7
testing.tRunner(0xc4215060f0, 0xc421564270)
	/usr/local/go/src/testing/testing.go:746 +0x16d
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:789 +0x569

goroutine 1 [chan receive]:
testing.(*T).Run(0xc4202e6690, 0x2954be8, 0x15, 0x2a1a080, 0xc420641c01)
	/usr/local/go/src/testing/testing.go:790 +0x59b
testing.runTests.func1(0xc4202e6690)
	/usr/local/go/src/testing/testing.go:1004 +0xa8
testing.tRunner(0xc4202e6690, 0xc420641d50)
	/usr/local/go/src/testing/testing.go:746 +0x16d
testing.runTests(0xc4204ceb40, 0x3ba0860, 0x34, 0x34, 0x38b21dc)
	/usr/local/go/src/testing/testing.go:1002 +0x522
testing.(*M).Run(0xc420a85f20, 0x18)
	/usr/local/go/src/testing/testing.go:921 +0x207
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun_test.TestMain(0xc420641f20)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/main_test.go:36 +0x15e
main.main()
	github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/_test/_testmain.go:168 +0x1ce

goroutine 38 [chan receive]:
github.com/cockroachdb/cockroach/pkg/util/log.(*loggingT).flushDaemon(0x3df3380)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:1053 +0x8f
created by github.com/cockroachdb/cockroach/pkg/util/log.init.0
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:581 +0x13a

goroutine 39 [syscall]:
os/signal.signal_recv(0x84b471)
	/usr/local/go/src/runtime/sigqueue.go:131 +0xa6
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:22 +0x30
created by os/signal.init.0
	/usr/local/go/src/os/signal/signal_unix.go:28 +0x4f

goroutine 29 [select, locked to thread]:
runtime.gopark(0x2a1d258, 0x0, 0x29404e5, 0x6, 0x18, 0x1)
	/usr/local/go/src/runtime/proc.go:287 +0xfd
runtime.selectgo(0xc42004bf50, 0xc4203281e0)
	/usr/local/go/src/runtime/select.go:395 +0x12c3
runtime.ensureSigM.func1()
	/usr/local/go/src/runtime/signal_unix.go:511 +0x1a7
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:2337 +0x1

goroutine 3353 [chan send]:
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.TestFlowRegistryDrain.func5.1(0xc421c32090, 0x989680, 0xc420ce62a0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow_registry_test.go:447 +0x6a
created by github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.TestFlowRegistryDrain.func5
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow_registry_test.go:445 +0xa7

goroutine 3348 [chan receive]:
testing.(*T).Run(0xc421506780, 0x2946ea4, 0xb, 0xc421564270, 0xc421c32001)
	/usr/local/go/src/testing/testing.go:790 +0x59b
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.TestFlowRegistryDrain(0xc421506780)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow_registry_test.go:441 +0x89c
testing.tRunner(0xc421506780, 0x2a1a080)
	/usr/local/go/src/testing/testing.go:746 +0x16d
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:789 +0x569


ERROR: exit status 2

context canceled
make: *** [stress] Error 1
Makefile:812: recipe for target 'stress' failed
```

--- a/pkg/sql/distsqlrun/flow_registry.go
+++ b/pkg/sql/distsqlrun/flow_registry.go
@@ -101,6 +101,11 @@ type flowRegistry struct {
 
 	// flowDone is signaled whenever the size of flows decreases.
 	flowDone *sync.Cond
+
+	// testingRunBeforeDrainSleep is a testing knob executed when a draining
+	// flowRegistry has no registered flows but must still wait for a minimum time
+	// for any incoming flows to register.
+	testingRunBeforeDrainSleep func()
 }
 
 // makeFlowRegistry creates a new flowRegistry.
@@ -301,6 +306,13 @@ func (fr *flowRegistry) Drain(flowDrainWait time.Duration, minFlowDrainWait time
 	start := timeutil.Now()
 	stopWaiting := false
 
+	sleep := func(t time.Duration) {
+		if fr.testingRunBeforeDrainSleep != nil {
+			fr.testingRunBeforeDrainSleep()
+		}
+		time.Sleep(t)
+	}
+
 	defer func() {
 		// At this stage, we have either hit the flowDrainWait timeout or we have no
 		// flows left. We wait for an expectedConnectionTime longer so that we give
@@ -323,7 +335,7 @@ func (fr *flowRegistry) Drain(flowDrainWait time.Duration, minFlowDrainWait time
 	fr.Lock()
 	if len(fr.flows) == 0 {
 		fr.Unlock()
-		time.Sleep(minFlowDrainWait)
+		sleep(minFlowDrainWait)
 		fr.Lock()
 		// No flows were registered, return.
 		if len(fr.flows) == 0 {
@@ -353,7 +365,7 @@ func (fr *flowRegistry) Drain(flowDrainWait time.Duration, minFlowDrainWait time
 	// finish.
 	waitTime := timeutil.Since(start)
 	if waitTime < minFlowDrainWait {
-		time.Sleep(minFlowDrainWait - waitTime)
+		sleep(minFlowDrainWait - waitTime)
 		fr.Lock()
 		for !(stopWaiting || len(fr.flows) == 0) {
 			fr.flowDone.Wait()

--- a/pkg/sql/distsqlrun/flow_registry_test.go
+++ b/pkg/sql/distsqlrun/flow_registry_test.go
@@ -24,6 +24,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/cockroachdb/cockroach/pkg/util/uuid"
 	"github.com/pkg/errors"
 )
@@ -373,7 +374,7 @@ func TestFlowRegistryDrain(t *testing.T) {
 	flow := &Flow{}
 	flow.Ctx = ctx
 	id := FlowID{uuid.MakeV4()}
-	registerFlow := func(id FlowID) {
+	registerFlow := func(t *testing.T, id FlowID) {
 		t.Helper()
 		if err := reg.RegisterFlow(
 			ctx, id, flow, nil /* inboundStreams */, 0, /* timeout */
@@ -385,7 +386,7 @@ func TestFlowRegistryDrain(t *testing.T) {
 	// WaitForFlow verifies that Drain waits for a flow to finish within the
 	// timeout.
 	t.Run("WaitForFlow", func(t *testing.T) {
-		registerFlow(id)
+		registerFlow(t, id)
 		drainDone := make(chan struct{})
 		go func() {
 			reg.Drain(math.MaxInt64 /* flowDrainWait */, 0 /* minFlowDrainWait */)
@@ -400,7 +401,7 @@ func TestFlowRegistryDrain(t *testing.T) {
 
 	// DrainTimeout verifies that Drain returns once the timeout expires.
 	t.Run("DrainTimeout", func(t *testing.T) {
-		registerFlow(id)
+		registerFlow(t, id)
 		reg.Drain(0 /* flowDrainWait */, 0 /* minFlowDrainWait */)
 		reg.UnregisterFlow(id)
 		reg.Undrain()
@@ -409,7 +410,7 @@ func TestFlowRegistryDrain(t *testing.T) {
 	// AcceptNewFlow verifies that a flowRegistry continues accepting flows
 	// while draining.
 	t.Run("AcceptNewFlow", func(t *testing.T) {
-		registerFlow(id)
+		registerFlow(t, id)
 		drainDone := make(chan struct{})
 		go func() {
 			reg.Drain(math.MaxInt64 /* flowDrainWait */, 0 /* minFlowDrainWait */)
@@ -418,7 +419,7 @@ func TestFlowRegistryDrain(t *testing.T) {
 		// Be relatively sure that the flowRegistry is draining.
 		time.Sleep(time.Microsecond)
 		newFlowID := FlowID{uuid.MakeV4()}
-		registerFlow(newFlowID)
+		registerFlow(t, newFlowID)
 		reg.UnregisterFlow(id)
 		select {
 		case <-drainDone:
@@ -439,22 +440,49 @@ func TestFlowRegistryDrain(t *testing.T) {
 	// MinFlowWait verifies that the flowRegistry waits a minimum amount of time
 	// for incoming flows to be registered.
 	t.Run("MinFlowWait", func(t *testing.T) {
-		minFlowDrainWait := 10 * time.Millisecond
 		// Case in which draining is initiated with zero running flows.
 		drainDone := make(chan struct{})
+		// Register a flow right before the flowRegistry waits for
+		// minFlowDrainWait. Use an errChan because draining is performed from
+		// another goroutine and cannot call t.Fatal.
+		errChan := make(chan error)
+		reg.testingRunBeforeDrainSleep = func() {
+			if err := reg.RegisterFlow(
+				ctx, id, flow, nil /* inboundStreams */, 0, /* timeout */
+			); err != nil {
+				errChan <- err
+			}
+			errChan <- nil
+		}
+		defer func() { reg.testingRunBeforeDrainSleep = nil }()
 		go func() {
-			reg.Drain(math.MaxInt64 /* flowDrainWait */, minFlowDrainWait)
+			reg.Drain(math.MaxInt64 /* flowDrainWait */, 0 /* minFlowDrainWait */)
 			drainDone <- struct{}{}
 		}()
-		// Be relatively sure that the flowRegistry is draining.
-		time.Sleep(time.Microsecond)
-		registerFlow(id)
+		if err := <-errChan; err != nil {
+			t.Fatal(err)
+		}
 		reg.UnregisterFlow(id)
 		<-drainDone
 		reg.Undrain()
 
-		// Case in which a running flow finishes before the minimum wait time.
-		registerFlow(id)
+		// Case in which a running flow finishes before the minimum wait time. We
+		// attempt to register another flow after the completion of the first flow
+		// to simulate an incoming flow that is registered during the minimum wait
+		// time. However, it is possible to unregister the first flow after the
+		// minimum wait time has passed, in which case we simply verify that the
+		// flowRegistry drain process has lasted at least the required wait time.
+		registerFlow(t, id)
+		reg.testingRunBeforeDrainSleep = func() {
+			if err := reg.RegisterFlow(
+				ctx, id, flow, nil /* inboundStreams */, 0, /* timeout */
+			); err != nil {
+				errChan <- err
+			}
+			errChan <- nil
+		}
+		minFlowDrainWait := 10 * time.Millisecond
+		start := timeutil.Now()
 		go func() {
 			reg.Drain(math.MaxInt64 /* flowDrainWait */, minFlowDrainWait)
 			drainDone <- struct{}{}
@@ -462,9 +490,20 @@ func TestFlowRegistryDrain(t *testing.T) {
 		// Be relatively sure that the flowRegistry is draining.
 		time.Sleep(time.Microsecond)
 		reg.UnregisterFlow(id)
-		registerFlow(id)
+		select {
+		case <-drainDone:
+			if timeutil.Since(start) < minFlowDrainWait {
+				t.Fatal("flow registry did not wait at least minFlowDrainWait")
+			}
+			return
+		case err := <-errChan:
+			if err != nil {
+				t.Fatal(err)
+			}
+		}
 		reg.UnregisterFlow(id)
 		<-drainDone
+
 		reg.Undrain()
 	})
 }

